File: src/EnterpriseECommerce.Application/Services/ProductService.cs
using EnterpriseECommerce.Application.DTOs;
using EnterpriseECommerce.Application.Interfaces;
using EnterpriseECommerce.Domain.Entities;

namespace EnterpriseECommerce.Application.Services;

/// <summary>
/// Contains product-related business logic.
///
/// Responsibilities:
/// - Retrieve active products
/// - Retrieve all products for Admin
/// - Retrieve low-stock products
/// - Create products
/// - Reactivate inactive products with the same SKU
/// - Update products
/// - Manage GST / HSN information
/// - Increase inventory stock
/// - Decrease inventory stock
/// - Soft-deactivate products
/// - Map Product entities to ProductDto
/// </summary>
public class ProductService
{
    private readonly IProductRepository
        _productRepository;

    public ProductService(
        IProductRepository productRepository)
    {
        _productRepository =
            productRepository;
    }

    // ============================================================
    // GET ALL ACTIVE PRODUCTS
    // ============================================================

    public async Task<IReadOnlyList<ProductDto>>
        GetAllProductsAsync()
    {
        var products =
            await _productRepository
                .GetAllAsync();

        return products
            .Where(
                product =>
                    product.IsActive)
            .Select(
                MapToDto)
            .ToList();
    }

    // ============================================================
    // GET PRODUCT BY ID
    // ============================================================

    public async Task<ProductDto?>
        GetProductByIdAsync(
            Guid id)
    {
        if (id ==
            Guid.Empty)
        {
            return null;
        }

        var product =
            await _productRepository
                .GetByIdAsync(
                    id);

        if (product is null ||
            !product.IsActive)
        {
            return null;
        }

        return MapToDto(
            product);
    }

    // ============================================================
    // CREATE PRODUCT
    // ============================================================

    public async Task<ProductDto>
        CreateProductAsync(
            CreateProductRequest request)
    {
        ArgumentNullException.ThrowIfNull(
            request);

        // ========================================================
        // VALIDATION
        // ========================================================

        if (request.CategoryId ==
            Guid.Empty)
        {
            throw new ArgumentException(
                "CategoryId is required.");
        }

        if (string.IsNullOrWhiteSpace(
            request.Name))
        {
            throw new ArgumentException(
                "Product name is required.");
        }

        if (string.IsNullOrWhiteSpace(
            request.SKU))
        {
            throw new ArgumentException(
                "SKU is required.");
        }

        if (request.Price < 0)
        {
            throw new ArgumentException(
                "Price cannot be negative.");
        }

        if (request.StockQuantity < 0)
        {
            throw new ArgumentException(
                "Stock quantity cannot be negative.");
        }

        ValidateGstRate(
            request.GstRate);

        // HSN is allowed to remain blank temporarily.
        // Existing products migrated from the previous model
        // will also initially have an empty HSN code.

        var normalizedName =
            request.Name.Trim();

        var normalizedDescription =
            request.Description?.Trim()
            ?? string.Empty;

        var normalizedSku =
            request.SKU
                .Trim()
                .ToUpperInvariant();

        var normalizedHsn =
            request.HsnCode?.Trim()
            ?? string.Empty;

        // ========================================================
        // CHECK EXISTING SKU
        // ========================================================

        var existingProduct =
            await _productRepository
                .GetBySkuAsync(
                    normalizedSku);

        if (existingProduct is not null)
        {
            // ====================================================
            // ACTIVE DUPLICATE
            // ====================================================

            if (existingProduct.IsActive)
            {
                throw new InvalidOperationException(
                    "A product with this SKU already exists.");
            }

            // ====================================================
            // REACTIVATE PREVIOUS PRODUCT
            // ====================================================

            existingProduct.UpdateCategory(
                request.CategoryId);

            existingProduct.UpdateDetails(
                normalizedName,
                normalizedDescription);

            existingProduct.UpdatePrice(
                request.Price);

            existingProduct.UpdateStock(
                request.StockQuantity);

            existingProduct.UpdateTaxInformation(
                normalizedHsn,
                request.GstRate);

            existingProduct.Activate();

            await _productRepository
                .UpdateAsync(
                    existingProduct);

            return MapToDto(
                existingProduct);
        }

        // ========================================================
        // CREATE NEW PRODUCT
        // ========================================================

        var product =
            new Product(
                categoryId:
                    request.CategoryId,

                name:
                    normalizedName,

                description:
                    normalizedDescription,

                sku:
                    normalizedSku,

                price:
                    request.Price,

                stockQuantity:
                    request.StockQuantity,

                hsnCode:
                    normalizedHsn,

                gstRate:
                    request.GstRate);

        await _productRepository
            .AddAsync(
                product);

        return MapToDto(
            product);
    }

    // ============================================================
    // UPDATE PRODUCT
    // ============================================================

    public async Task<ProductDto?>
        UpdateProductAsync(
            Guid id,
            UpdateProductRequest request)
    {
        ArgumentNullException.ThrowIfNull(
            request);

        if (id ==
            Guid.Empty)
        {
            return null;
        }

        if (string.IsNullOrWhiteSpace(
            request.Name))
        {
            throw new ArgumentException(
                "Product name is required.");
        }

        if (request.Price < 0)
        {
            throw new ArgumentException(
                "Price cannot be negative.");
        }

        if (request.StockQuantity < 0)
        {
            throw new ArgumentException(
                "Stock quantity cannot be negative.");
        }

        ValidateGstRate(
            request.GstRate);

        var product =
            await _productRepository
                .GetByIdAsync(
                    id);

        if (product is null ||
            !product.IsActive)
        {
            return null;
        }

        product.UpdateDetails(
            request.Name.Trim(),
            request.Description?.Trim()
            ?? string.Empty);

        product.UpdatePrice(
            request.Price);

        product.UpdateStock(
            request.StockQuantity);

        product.UpdateTaxInformation(
            request.HsnCode,
            request.GstRate);

        await _productRepository
            .UpdateAsync(
                product);

        return MapToDto(
            product);
    }

    // ============================================================
    // ADMIN - INCREASE STOCK
    // ============================================================

    public async Task<ProductDto>
        IncreaseProductStockAsync(
            Guid productId,
            int quantity)
    {
        if (productId ==
            Guid.Empty)
        {
            throw new ArgumentException(
                "ProductId is required.");
        }

        if (quantity <= 0)
        {
            throw new ArgumentException(
                "Quantity must be greater than zero.");
        }

        var product =
            await _productRepository
                .GetByIdAsync(
                    productId);

        if (product is null ||
            !product.IsActive)
        {
            throw new KeyNotFoundException(
                "Product not found.");
        }

        product.IncreaseStock(
            quantity);

        await _productRepository
            .UpdateAsync(
                product);

        return MapToDto(
            product);
    }

    // ============================================================
    // ADMIN - DECREASE STOCK
    // ============================================================

    public async Task<ProductDto>
        DecreaseProductStockAsync(
            Guid productId,
            int quantity)
    {
        if (productId ==
            Guid.Empty)
        {
            throw new ArgumentException(
                "ProductId is required.");
        }

        if (quantity <= 0)
        {
            throw new ArgumentException(
                "Quantity must be greater than zero.");
        }

        var product =
            await _productRepository
                .GetByIdAsync(
                    productId);

        if (product is null ||
            !product.IsActive)
        {
            throw new KeyNotFoundException(
                "Product not found.");
        }

        product.ReduceStock(
            quantity);

        await _productRepository
            .UpdateAsync(
                product);

        return MapToDto(
            product);
    }

    // ============================================================
    // DEACTIVATE PRODUCT
    // ============================================================

    public async Task<bool>
        DeactivateProductAsync(
            Guid id)
    {
        if (id ==
            Guid.Empty)
        {
            return false;
        }

        var product =
            await _productRepository
                .GetByIdAsync(
                    id);

        if (product is null ||
            !product.IsActive)
        {
            return false;
        }

        product.Deactivate();

        await _productRepository
            .UpdateAsync(
                product);

        return true;
    }

    // ============================================================
    // ADMIN - GET ALL PRODUCTS
    // ============================================================

    public async Task<IReadOnlyList<ProductDto>>
        GetAllProductsForAdminAsync()
    {
        var products =
            await _productRepository
                .GetAllAsync();

        return products
            .Select(
                MapToDto)
            .ToList();
    }

    // ============================================================
    // ADMIN - GET LOW STOCK PRODUCTS
    // ============================================================

    public async Task<IReadOnlyList<ProductDto>>
        GetLowStockProductsAsync(
            int threshold = 5)
    {
        if (threshold < 0)
        {
            throw new ArgumentException(
                "Low stock threshold cannot be negative.");
        }

        var products =
            await _productRepository
                .GetAllAsync();

        return products
            .Where(
                product =>
                    product.IsActive &&
                    product.StockQuantity <=
                    threshold)
            .OrderBy(
                product =>
                    product.StockQuantity)
            .Select(
                MapToDto)
            .ToList();
    }

    // ============================================================
    // VALIDATE GST
    // ============================================================

    private static void ValidateGstRate(
        decimal gstRate)
    {
        if (gstRate < 0 ||
            gstRate > 100)
        {
            throw new ArgumentException(
                "GST rate must be between 0 and 100.");
        }
    }

    // ============================================================
    // ENTITY -> DTO
    // ============================================================

    private static ProductDto MapToDto(
        Product product)
    {
        return new ProductDto
        {
            Id =
                product.Id,

            CategoryId =
                product.CategoryId,

            Name =
                product.Name,

            Description =
                product.Description,

            SKU =
                product.SKU,

            HsnCode =
                product.HsnCode,

            GstRate =
                product.GstRate,

            Price =
                product.Price,

            StockQuantity =
                product.StockQuantity,

            IsActive =
                product.IsActive,

            CreatedAt =
                product.CreatedAt,

            UpdatedAt =
                product.UpdatedAt
        };
    }
}

File: client/src/services/productService.js
import apiClient from '../api/apiClient'

// ============================================================
// PRODUCT SERVICE
// ============================================================
//
// Handles customer and admin product API calls.
//
// Customer:
// - View active products
//
// Admin:
// - View all products
// - View low-stock products
// - Create product
// - Update product
// - Deactivate product
// ============================================================

// ------------------------------------------------------------
// GET ACTIVE PRODUCTS
// ------------------------------------------------------------
//
// GET /api/Products
// ------------------------------------------------------------

export async function getProducts() {

    const response =
        await apiClient.get(
            '/Products'
        )

    return response.data
}

// ------------------------------------------------------------
// GET PRODUCT BY ID
// ------------------------------------------------------------
//
// GET /api/Products/{id}
// ------------------------------------------------------------

export async function getProductById(
    productId
) {

    const response =
        await apiClient.get(
            `/Products/${productId}`
        )

    return response.data
}

// ------------------------------------------------------------
// ADMIN - GET ALL PRODUCTS
// ------------------------------------------------------------
//
// GET /api/Products/admin/all
//
// Includes:
// - Active products
// - Inactive products
// ------------------------------------------------------------

export async function getAllProductsForAdmin() {

    const response =
        await apiClient.get(
            '/Products/admin/all'
        )

    return response.data
}

// ------------------------------------------------------------
// ADMIN - GET LOW STOCK PRODUCTS
// ------------------------------------------------------------
//
// GET /api/Products/admin/low-stock
//
// Optional:
//
// GET /api/Products/admin/low-stock?threshold=10
//
// Default threshold is 5.
// ------------------------------------------------------------

export async function getLowStockProducts(
    threshold = 5
) {

    const response =
        await apiClient.get(
            '/Products/admin/low-stock',
            {
                params: {
                    threshold
                }
            }
        )

    return response.data
}

// ------------------------------------------------------------
// CREATE PRODUCT
// ------------------------------------------------------------
//
// POST /api/Products
//
// Admin only.
// ------------------------------------------------------------

export async function createProduct(
    productData
) {

    const response =
        await apiClient.post(
            '/Products',
            productData
        )

    return response.data
}

// ------------------------------------------------------------
// UPDATE PRODUCT
// ------------------------------------------------------------
//
// PUT /api/Products/{id}
//
// Admin only.
// ------------------------------------------------------------

export async function updateProduct(
    productId,
    productData
) {

    const response =
        await apiClient.put(
            `/Products/${productId}`,
            productData
        )

    return response.data
}

// ------------------------------------------------------------
// ADMIN - INCREASE PRODUCT STOCK
// ------------------------------------------------------------
//
// POST /api/Products/{id}/stock/increase
//
// Request:
// {
//   quantity: 5
// }
// ------------------------------------------------------------

export async function increaseProductStock(
    productId,
    quantity
) {

    const response =
        await apiClient.post(
            `/Products/${productId}/stock/increase`,
            {
                quantity
            }
        )

    return response.data
}

// ------------------------------------------------------------
// ADMIN - DECREASE PRODUCT STOCK
// ------------------------------------------------------------
//
// POST /api/Products/{id}/stock/decrease
// ------------------------------------------------------------

export async function decreaseProductStock(
    productId,
    quantity
) {

    const response =
        await apiClient.post(
            `/Products/${productId}/stock/decrease`,
            {
                quantity
            }
        )

    return response.data
}

// ------------------------------------------------------------
// DEACTIVATE PRODUCT
// ------------------------------------------------------------
//
// DELETE /api/Products/{id}
//
// Admin only.
//
// This is a soft delete.
// ------------------------------------------------------------

export async function deactivateProduct(
    productId
) {

    const response =
        await apiClient.delete(
            `/Products/${productId}`
        )

    return response.data
}